File: src/main/java/net/atos/client/zgw/shared/exception/ZgwValidationErrorException.java
/*
 * SPDX-FileCopyrightText: 2021 Atos, 2025 INFO.nl
 * SPDX-License-Identifier: EUPL-1.2+
 */
package net.atos.client.zgw.shared.exception;

import java.net.URI;
import java.util.stream.Collectors;

import net.atos.client.zgw.shared.model.ValidationZgwError;
import nl.info.zac.exception.InputValidationFailedException;

/**
 * Exception to indicate a validation error that occurred in when calling the ZGW API.
 */
public class ZgwValidationErrorException extends InputValidationFailedException {

    private final ValidationZgwError validatieFout;

    public ZgwValidationErrorException(final ValidationZgwError validationZgwError) {
        this.validatieFout = validationZgwError;
    }

    public ValidationZgwError getValidatieFout() {
        return validatieFout;
    }

    @Override
    public String getMessage() {
        return "%s [%d %s] %s: %s (%s %s)"
                .formatted(validatieFout.getTitle(),
                        validatieFout.getStatus(),
                        validatieFout.getCode(),
                        validatieFout.getDetail(),
                        validatieFout.getInvalidParams().stream()
                                .map(error -> "%s [%s] %s"
                                        .formatted(
                                                error.name(),
                                                error.code(),
                                                error.reason()
                                        )
                                )
                                .collect(Collectors.joining(", ")),
                        uri(validatieFout.getType()),
                        uri(validatieFout.getInstance()));
    }

    private String uri(final URI uri) {
        return uri == null ? null : uri.toString();
    }
}

File: src/main/kotlin/nl/info/client/zgw/shared/exception/ZgwValidationErrorException.kt
/*
 * SPDX-FileCopyrightText: 2021 Atos, 2025 INFO.nl
 * SPDX-License-Identifier: EUPL-1.2+
 */
package nl.info.client.zgw.shared.exception

import nl.info.client.zgw.shared.model.ZgwValidationError
import nl.info.zac.exception.InputValidationFailedException

/**
 * Exception to indicate a validation error that occurred in when calling the ZGW API.
 */
class ZgwValidationErrorException(val validatieFout: ZgwValidationError) : InputValidationFailedException() {

    override val message: String
        get() = "%s [%d %s] %s: %s (%s %s)".format(
            validatieFout.title,
            validatieFout.status,
            validatieFout.code,
            validatieFout.detail,
            validatieFout.invalidParams.joinToString(", ") { "${it.name} [${it.code}] ${it.reason}" },
            validatieFout.type,
            validatieFout.instance
        )
}
